ext/random/engine_mt19937.c

```diff
@@ -0,0 +1,343 @@
+#define hiBit(u)      ((u) & 0x80000000U)  /* mask all but highest   bit of u */
+#define loBit(u)      ((u) & 0x00000001U)  /* mask all but lowest    bit of u */
+#define loBits(u)     ((u) & 0x7FFFFFFFU)  /* mask     the highest   bit of u */
+#define mixBits(u, v) (hiBit(u) | loBits(v)) /* move hi bit of u to hi bit of v */
+
+#define twist(m,u,v)  (m ^ (mixBits(u,v) >> 1) ^ ((uint32_t)(-(int32_t)(loBit(v))) & 0x9908b0dfU))
+#define twist_php(m,u,v)  (m ^ (mixBits(u,v) >> 1) ^ ((uint32_t)(-(int32_t)(loBit(u))) & 0x9908b0dfU))
+
+static inline void mt19937_reload(php_random_status_state_mt19937 *state) {
+	uint32_t *p = state->state;
+	int i;
+
+	if (state->mode == MT_RAND_MT19937) {
+		for (i = N - M; i--; ++p) {
+			*p = twist(p[M], p[0], p[1]);
+		}
+		for (i = M; --i; ++p) {
+			*p = twist(p[M-N], p[0], p[1]);
+		}
+		*p = twist(p[M-N], p[0], state->state[0]);
+	} else {
+		for (i = N - M; i--; ++p) {
+			*p = twist_php(p[M], p[0], p[1]);
+		}
+		for (i = M; --i; ++p) {
+			*p = twist_php(p[M-N], p[0], p[1]);
+		}
+		*p = twist_php(p[M-N], p[0], state->state[0]);
+	}
+
+	state->count = 0;
+}
+
+static inline void mt19937_seed_state(php_random_status_state_mt19937 *state, uint64_t seed)
+{
+	uint32_t i, prev_state;
+
+	/* Initialize generator state with seed
+	   See Knuth TAOCP Vol 2, 3rd Ed, p.106 for multiplier.
+	   In previous versions, most significant bits (MSBs) of the seed affect
+	   only MSBs of the state array.  Modified 9 Jan 2002 by Makoto Matsumoto. */
+	state->state[0] = seed & 0xffffffffU;
+	for (i = 1; i < MT_N; i++) {
+		prev_state = state->state[i - 1];
+		state->state[i] = (1812433253U * (prev_state  ^ (prev_state  >> 30)) + i) & 0xffffffffU;
+	}
+	state->count = i;
+
+	mt19937_reload(state);
+}
+
+static void seed(php_random_status *status, uint64_t seed)
+{
+	mt19937_seed_state(status->state, seed);
+}
+
+static uint64_t generate(php_random_status *status)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	uint32_t s1;
+
+	if (s->count >= MT_N) {
+		mt19937_reload(s);
+	}
+
+	s1 = s->state[s->count++];
+	s1 ^= (s1 >> 11);
+	s1 ^= (s1 << 7) & 0x9d2c5680U;
+	s1 ^= (s1 << 15) & 0xefc60000U;
+
+	return (uint64_t) (s1 ^ (s1 >> 18));
+}
+
+static zend_long range(php_random_status *status, zend_long min, zend_long max)
+{
+	php_random_status_state_mt19937 *s = status->state;
+
+	if (s->mode == MT_RAND_MT19937) {
+		return php_random_range(&php_random_algo_mt19937, status, min, max);
+	}
+
+	uint64_t r = php_random_algo_mt19937.generate(status) >> 1;
+	/* Legacy mode deliberately not inside php_mt_rand_range()
+	 * to prevent other functions being affected */
+	RAND_RANGE_BADSCALING(r, min, max, PHP_MT_RAND_MAX);
+	return (zend_long) r;
+}
+
+static bool serialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_mt19937 *s = status->state;
+	zval tmp;
+	uint32_t i;
+
+	for (i = 0; i< MT_N; i++) {
+		ZVAL_LONG(&tmp, s->state[i]);
+		zend_hash_next_index_insert(data, &tmp);
+	}
```

Comment:
I think this should also be hexadecimal for consistency (e.g. for 32 bit platforms).

Reply:
fixed e5ff0510842c35e04a554e3e506423d8fed04cca